### PR TITLE
Always public IAM endpoint for satellite cluster

### PR DESCRIPTION
**Pull request by GunaKKIBM**

### CODE REVIEW

File: pkg/config/cluster_config.go [hunk 1]
```diff
@@ -110,28 +113,71 @@ func GetTokenExchangeURLfromStorageSecretStore(config Config, providerType strin
...
+	if err != nil {
+		return false, err
+	}
+
+	if len(nodeList.Items) == 0 {
+		return false, utils.Error{Description: "No nodes found"}
+	}
+
+	for _, node := range nodeList.Items {
+		providerType := node.ObjectMeta.Labels[utils.IaasProviderNodeLabel]
```
[the remark this answers is not in the document]
Comment (GunaKKIBM) [reply pkg/config/cluster_config.go:176]: Based on the node spec I have observed, there is no other way for the application to determine if the cluster is satellite, and even storage metrics agent is checking node labels to determine if the cluster is satellite. And max number of nodes is not going to cause any issue, it is going to list maximum of 3 nodes, generally, even if the cluster has less number of nodes, it shouldn't cause any issue.

File: pkg/config/cluster_config.go [hunk 2]
```diff
@@ -110,28 +113,71 @@ func GetTokenExchangeURLfromStorageSecretStore(config Config, providerType strin
...
-	// If the cluster is not satellite cluster, use PROD or STAGE URLs
-	if os.Getenv("IS_SATELLITE") != constTrue {
-		if !strings.Contains(url, "stage") && !strings.Contains(url, "test") {
-			url = utils.ProdPrivateIAMURL
-		} else {
-			url = utils.StagePrivateIAMURL
-		}
+	isSatellite, err := isSatellite(kc)
+	if err != nil {
+		return "", utils.Error{Description: "Unable to determine IAAS provider type", BackendError: err.Error()}
```
[the remark this answers is not in the document]
Comment (GunaKKIBM) [reply pkg/config/cluster_config.go:118]: I can change the error message. But, utils.Error is something that implements Golang error. And nothing is getting logged here.

File: pkg/config/cluster_config.go [hunk 3]
```diff
@@ -110,28 +113,71 @@ func GetTokenExchangeURLfromStorageSecretStore(config Config, providerType strin
...
 	}
 
-	// If the cluster is not satellite cluster, use PROD or STAGE URLs
-	if os.Getenv("IS_SATELLITE") != constTrue {
-		if !strings.Contains(url, "stage") && !strings.Contains(url, "test") {
-			url = utils.ProdPrivateIAMURL
-		} else {
-			url = utils.StagePrivateIAMURL
-		}
+	isSatellite, err := isSatellite(kc)
```
[the remark this answers is not in the document]
Comment (GunaKKIBM) [reply pkg/config/cluster_config.go:116]: This, I will check for a better way to do

[the remark this answers is not in the document]
Comment (GunaKKIBM) [reply pkg/config/cluster_config.go:116]: https://github.com/IBM/secret-utils-lib/blob/master/pkg/config/secret_config.go#L28 - Config is like this, which is very much adherent to slclient.toml structure, hence won't be adding it here. But, I will move out isSatellite function.

File: pkg/config/cluster_config.go [hunk 4]
```diff
@@ -72,29 +74,31 @@ func FrameTokenExchangeURL(kc k8s_utils.KubernetesClient, providerType string, l
 		return cloudConf.TokenExchangeURL + tokenExchangePath
 	}
 
+	cc, err := GetClusterInfo(kc, logger)
+	if err != nil {
+		logger.Error("Error fetching cluster info", zap.Error(err))
+		return (utils.ProdPublicIAMURL + tokenExchangePath)
```
Comment (GunaKKIBM) [inline pkg/config/cluster_config.go:80]: Unable to determine satellite/not, fall back to private IAM URL

File: pkg/config/cluster_config.go [hunk 5]
```diff
@@ -37,12 +36,15 @@ const (
 	tokenExchangePath = "/identity/token"
 	// constTrue ...
 	constTrue = "True"
+	// maxNodes ...
+	maxNodes = 3
```
Comment (GunaKKIBM) [inline pkg/config/cluster_config.go:40]: Remove this - not necessary

File: pkg/config/cluster_config.go [hunk 6]
```diff
@@ -110,28 +114,53 @@ func GetTokenExchangeURLfromStorageSecretStore(config Config, providerType strin
...
+// isProduction determines if the env in which a pod is deployed is stage or production
+func isProduction(url string) bool {
+	if !strings.Contains(url, "stage") && !strings.Contains(url, "test") {
+		return true
+	}
+	return false
+}
+
+// IsSatellite checks if the cluster where the pod is currently running is a satellite cluster or not
+func IsSatellite(cc ClusterConfig, logger *zap.Logger) bool {
```
Comment (GunaKKIBM) [inline pkg/config/cluster_config.go:160]: Check if configmap cluster-info is editable

Comment (arahamad) [reply pkg/config/cluster_config.go:154]: going forward lets plan to use get cluster type from immutable object

**Review by arahamad — APPROVED** [review]
lgtm, please verify on stage

**Review by ambiknai — APPROVED** [review]
lgtm

**Review by skorati — APPROVED** [review]

**Review by sameshai — APPROVED** [review]

**Pull request merged**